sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs

```diff
@@ -0,0 +1,154 @@
+                    }
+
+                    var enclosingMethod = simpleMemberAccess.FirstAncestorOrSelf<BaseMethodDeclarationSyntax>();
+                    if (enclosingMethod != null)
+                    {
+                        if (memberAccessNameName == "Sleep" &&
+                            !enclosingMethod.Modifiers.Any(m => m.IsKind(SyntaxKind.AsyncKeyword)))
+                        {
+                            return; // Thread.Sleep should not be used only in async methods
+                        }
+
+                        var methodSymbol = c.SemanticModel.GetDeclaredSymbol(enclosingMethod);
+                        if (methodSymbol != null &&
+                            methodSymbol.IsMainMethod())
+                        {
+                            return; // Main methods are not subject to deadlock issue so no need to report an issue
+                        }
+                    }
+
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, simpleMemberAccess.GetLocation(),
+                        messageArgs: GetMessageArguments(memberAccessNameName)));
+                },
+                SyntaxKind.SimpleMemberAccessExpression);
+        }
+
+        private static bool IsPartOfDeadlockPreventingConstruct(MemberAccessExpressionSyntax memberAccess, SemanticModel semanticModel) =>
+            memberAccess.Expression.DescendantNodes()
+                .OfType<MemberAccessExpressionSyntax>()
+                .Any(subMemberAccess => IsDeadlockPreventingConstruct(subMemberAccess, semanticModel));
+
+        private static bool IsDeadlockPreventingConstruct(MemberAccessExpressionSyntax memberAccess, SemanticModel semanticModel)
+        {
+            if (memberAccess?.Name == null ||
+                !DeadlockPreventingConstructs.ContainsKey(memberAccess.Name.Identifier.ValueText))
+            {
+                return false;
+            }
+
+            var memberAccessSymbol = semanticModel.GetSymbolInfo(memberAccess).Symbol?.ContainingType?.ConstructedFrom;
+            return memberAccessSymbol.Is(DeadlockPreventingConstructs[memberAccess.Name.Identifier.ValueText]);
+        }
+
+        private static string[] GetMessageArguments(string memberAccessName)
+        {
+            switch (memberAccessName)
+            {
+                case "Result":
+                    return new[] { "Task.Result", "await" };
+                case "Wait":
+                    return new[] { "Task.Wait", "await" };
+                case "GetResult":
+                    return new[] { "Task.GetAwaiter.GetResult", "await" };
+                case "WaitAny":
+                    return new[] { "Task.WaitAny", "await Task.WhenAny" };
+                case "WaitAll":
+                    return new[] { "Task.WaitAll", "await Task.WhenAll" };
+                case "Sleep":
+                    return new[] { "Thread.Sleep", "await Task.Delay" };
+                default:
+                    return new string[0];
```

Comment:
I think the `default` clause will throw if executed (string.Format expects two arguments, we provide only one); if it is not needed perhaps we should remove it, or explicitly throw...